File: test_parse.py
import struct
with open('tests/fixtures/Pokemon Blaze Black v3.1 - Complete.sav', 'rb') as f:
    data = bytearray(f.read())

activeSlot = 0

partyCount = data[activeSlot + 0x18E04]
print('partyCount:', partyCount)

validPokemon = 0
for i in range(partyCount):
    offset = activeSlot + 0x18E08 + (i * 220)
    expectedChecksum = struct.unpack('<H', data[offset+6:offset+8])[0]
    
    prngSeed = expectedChecksum
    decryptedData = bytearray(128)
    
    for j in range(0, 128, 2):
        prngSeed = (prngSeed * 0x41C64E6D + 0x6073) & 0xFFFFFFFF
        prngWord = prngSeed >> 16
        encryptedWord = struct.unpack('<H', data[offset+8+j:offset+8+j+2])[0]
        decryptedWord = encryptedWord ^ prngWord
        struct.pack_into('<H', decryptedData, j, decryptedWord)
        
    actualChecksum = 0
    for j in range(0, 128, 2):
        actualChecksum = (actualChecksum + struct.unpack('<H', decryptedData[j:j+2])[0]) & 0xFFFF
        
    print(f'Slot {i} checksum: actual={actualChecksum} expected={expectedChecksum}')
    if actualChecksum == expectedChecksum:
        validPokemon += 1
        
print('validPokemon:', validPokemon)

File: test_parse.js
const fs = require('fs');

const data = fs.readFileSync('tests/fixtures/Pokemon Blaze Black v3.1 - Complete.sav');
const buffer = data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength);

// simulate what Gen5SaveParser does
const view = new DataView(buffer);
const activeSlot = 0; // assuming 0

const partyCount = view.getUint8(activeSlot + 0x18E04);
console.log('partyCount:', partyCount);

let validPokemon = 0;
for(let i=0; i<partyCount; i++) {
    const offset = activeSlot + 0x18E08 + (i * 220);
    const expectedChecksum = view.getUint16(offset + 6, true);
    let actualChecksum = 0;
    
    let prngSeed = expectedChecksum;
    const decryptedData = new Uint8Array(128);
    const decView = new DataView(decryptedData.buffer);
    
    for (let j = 0; j < 128; j += 2) {
      prngSeed = (Math.imul(prngSeed, 0x41C64E6D) + 0x6073) >>> 0;
      const prngWord = prngSeed >>> 16;
      const encryptedWord = view.getUint16(offset + 8 + j, true);
      const decryptedWord = encryptedWord ^ prngWord;
      decView.setUint16(j, decryptedWord, true);
    }
    
    for (let j = 0; j < 128; j += 2) {
      actualChecksum = (actualChecksum + decView.getUint16(j, true)) & 0xFFFF;
    }
    
    console.log('Slot ' + i + ' checksum:', 'actual=' + actualChecksum, 'expected=' + expectedChecksum);
    if(actualChecksum === expectedChecksum) {
        validPokemon++;
    }
}
console.log('validPokemon:', validPokemon);
